### ISSUE TITLE
MapView#snapshot returns a black bitmap

### ISSUE THREAD

**Issue by wingkingbobo** [mapbox/mapbox-maps-android#923]
## Environment

This problem is raised in issue #42, And it has been repaired.
But I still have this bug, and it occurs every time.
Do I need an updated version?
```
  implementation 'com.mapbox.maps:android:10.1.0'
```
```
        btn_snapshot.setOnClickListener {
            mMapView?.snapshot {
                val bitmap = it
                runOnUiThread {
                    image_view.setImageBitmap(bitmap)
                }
            }
//            thread {
//                val bitmap = mMapView?.snapshot()
//                runOnUiThread {
//                    image_view.setImageBitmap(bitmap)
//                }
//            }
        }
```

**Comment by kiryldz** [comment]
@wingkingbobo I've checked that everything should work fine, we have similar test [activity](https://github.com/mapbox/mapbox-maps-android/blob/main/app/src/main/java/com/mapbox/maps/testapp/examples/snapshotter/MapViewSnapshotActivity.kt). It uses sync method to get snapshot but I've validated that async (as you're using in your code snippet) also works fine.

Can you provide more info:
1. device you're using
2. your layout file that includes `MapView`
3. are you sure you're performing snapshot when `MapView` is not in paused / stopped state? activity or fragment holding `MapView` should be in `onResume` state.
4. any suspicious logs when you perform snapshot.

**Comment by wingkingbobo** [comment]
> @wingkingbobo I've checked that everything should work fine, we have similar test [activity](https://github.com/mapbox/mapbox-maps-android/blob/main/app/src/main/java/com/mapbox/maps/testapp/examples/snapshotter/MapViewSnapshotActivity.kt). It uses sync method to get snapshot but I've validated that async (as you're using in your code snippet) also works fine.
> 
> Can you provide more info:
> 
> 1. device you're using
> 2. your layout file that includes `MapView`
> 3. are you sure you're performing snapshot when `MapView` is not in paused / stopped state? activity or fragment holding `MapView` should be in `onResume` state.
> 4. any suspicious logs when you perform snapshot.

Source Code:
[TestSnapshot.zip](https://github.com/mapbox/mapbox-maps-android/files/7630325/TestSnapshot.zip)
mapbox_access_token in strings.xml needs to be replaced.

```kotlin
package com.wingkingbobo.testsnapshot

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import com.mapbox.maps.Style
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        map_view.getMapboxMap().loadStyleUri(Style.SATELLITE_STREETS) {
            btn_snapshot.setOnClickListener {

                map_view.snapshot { bitmap ->
                    runOnUiThread {
                        image_view.setImageBitmap(bitmap)
                    }
                }
            }
        }
    }
}
```

```xml
<?xml version="1.0" encoding="utf-8"?>
<FrameLayout xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:app="http://schemas.android.com/apk/res-auto"
    xmlns:tools="http://schemas.android.com/tools"
    android:layout_width="match_parent"
    android:layout_height="match_parent"
    tools:context=".MainActivity">

    <com.mapbox.maps.MapView
        android:id="@+id/map_view"
        android:layout_width="match_parent"
        android:layout_height="match_parent"/>

    <Button
        android:id="@+id/btn_snapshot"
        android:layout_width="wrap_content"
        android:layout_height="wrap_content"
        android:text="snapshot"
        android:layout_gravity="end" />

    <ImageView
        android:id="@+id/image_view"
        android:layout_width="100dp"
        android:layout_height="100dp"
        android:background="#fff"/>

</FrameLayout>
```

Here are the phones I tested. 
It seems that some Huawei mobile phone have problems.

| Phone                  | OS            | Snapshot Available |
| ---------------------- | ------------- | ------------------ |
| HONOR 20 PRO/ YAL-AL10 | Android 10    | no                 |
| HUAWEI nova 3e         | Android 9     | no                 |
| HUAWEI P30             | HarmonyOS 2.0 | no                 |
| HUAWEI nova 7          | HarmonyOS 2.0 | yes                |
| VIVO Y53s              | Android 11    | yes                |
| Redmi Note 9           | Android 10    | yes                |


All logs from app startup to snapshot call:
```
2021-12-01 11:04:48.036 25560-25560/? I/bo.testsnapsho: Late-enabling -Xcheck:jni
2021-12-01 11:04:48.063 25560-25560/? E/bo.testsnapsho: Unknown bits set in runtime_flags: 0x8000
2021-12-01 11:04:48.064 25560-25560/? I/bo.testsnapsho: Reinit property: dalvik.vm.checkjni= false
2021-12-01 11:04:48.064 25560-25560/? I/bo.testsnapsho: ReInitProperties: persist.vm.debug.dumpapi= false
2021-12-01 11:04:48.074 25560-25560/? I/ActivityThread: preloadStatus:1
2021-12-01 11:04:48.076 25560-25560/? D/ActivityThread: Attach thread to application
2021-12-01 11:04:48.085 25560-25578/com.wingkingbobo.testsnapshot I/HwActivityThreadImpl: mSocketName: 02a57a21-9431-4208-a5a2-db89b70c0137
2021-12-01 11:04:48.086 25560-25578/com.wingkingbobo.testsnapshot I/HwActivityThreadImpl: initPreloadedSocked end mPreloadStatus 2
2021-12-01 11:04:48.092 25560-25579/com.wingkingbobo.testsnapshot I/ActivityThread: Init compatible state: true
2021-12-01 11:04:48.093 25560-25579/com.wingkingbobo.testsnapshot D/HwFrameworkSecurityPartsFactory: HwFrameworkSecurityPartsFactory in.
2021-12-01 11:04:48.093 25560-25579/com.wingkingbobo.testsnapshot I/HwFrameworkSecurityPartsFactory: add HwFrameworkSecurityPartsFactory to memory.
2021-12-01 11:04:48.094 25560-25579/com.wingkingbobo.testsnapshot D/BehaviorCollectManager: hook process for AI virus second
2021-12-01 11:04:48.095 25560-25579/com.wingkingbobo.testsnapshot I/BehaviorCollectManager: bindAnalyzerService: huawei.android.security.IAppBehaviorDataAnalyzer$Stub$Proxy@6ca1d4d pid:25560
2021-12-01 11:04:48.102 25560-25580/com.wingkingbobo.testsnapshot I/HwActivityThreadImpl: before connect app.socketName:02a57a21-9431-4208-a5a2-db89b70c0137
2021-12-01 11:04:48.286 25560-25560/com.wingkingbobo.testsnapshot I/HwActivityThreadImpl: accept socket mSocketName: 02a57a21-9431-4208-a5a2-db89b70c0137
2021-12-01 11:04:48.286 25560-25560/com.wingkingbobo.testsnapshot I/HwActivityThreadImpl: accept socket end
2021-12-01 11:04:48.286 25560-25580/com.wingkingbobo.testsnapshot W/HwActivityThreadImpl: closeStream error.
2021-12-01 11:04:48.286 25560-25560/com.wingkingbobo.testsnapshot W/HwActivityThreadImpl: closeStream error.
2021-12-01 11:04:48.286 25560-25580/com.wingkingbobo.testsnapshot I/HwActivityThreadImpl: connectContinuePreload end mPreloadStatus 3
2021-12-01 11:04:48.286 25560-25560/com.wingkingbobo.testsnapshot I/HwActivityThreadImpl: acceptPreloadedApplication end mPreloadStatus 3
2021-12-01 11:04:48.288 25560-25579/com.wingkingbobo.testsnapshot I/HwApiCacheMangerEx: apicache path=/storage/emulated/0 state=mounted key=com.wingkingbobo.testsnapshot#11006#256
2021-12-01 11:04:48.288 25560-25579/com.wingkingbobo.testsnapshot I/HwApiCacheMangerEx: need clear apicache,because volumes changed,oldCnt=0 newCnt=1
2021-12-01 11:04:48.288 25560-25579/com.wingkingbobo.testsnapshot I/HwApiCacheMangerEx: apicache path=/storage/emulated/0 state=mounted key=com.wingkingbobo.testsnapshot#11006#0
2021-12-01 11:04:48.288 25560-25579/com.wingkingbobo.testsnapshot I/HwApiCacheMangerEx: need clear apicache,because volumes changed,oldCnt=1 newCnt=1
2021-12-01 11:04:48.293 25560-25579/com.wingkingbobo.testsnapshot I/HwPartIawareFactory: add android.common.HwPartIawareFactoryImpl to memory.
2021-12-01 11:04:48.293 25560-25579/com.wingkingbobo.testsnapshot I/AwareBitmapCacher: init processName:com.wingkingbobo.testsnapshot pid=25560 uid=11006
2021-12-01 11:04:48.294 25560-25600/com.wingkingbobo.testsnapshot E/AwareLog: AtomicFileUtils: readFileLines file not exist: android.util.AtomicFile@810ec7c
2021-12-01 11:04:48.311 25560-25560/com.wingkingbobo.testsnapshot I/MapboxCommonSDK: Loaded
2021-12-01 11:04:48.314 25560-25560/com.wingkingbobo.testsnapshot I/ActivityThread: finishPreloaded preloadStatus 3
2021-12-01 11:04:48.314 25560-25560/com.wingkingbobo.testsnapshot I/ActivityThread: finishPreloaded end preloadStatus 3
2021-12-01 11:04:48.314 25560-25560/com.wingkingbobo.testsnapshot W/ActivityThread: handleWindowVisibility: no activity for token android.os.BinderProxy@9834526
2021-12-01 11:04:48.324 25560-25560/com.wingkingbobo.testsnapshot D/OpenGLRenderer: disableOutlineDraw is true
2021-12-01 11:04:48.362 25560-25560/com.wingkingbobo.testsnapshot V/ActivityThread: callActivityOnCreate
2021-12-01 11:04:48.374 25560-25560/com.wingkingbobo.testsnapshot I/DecorView[]:  old windowMode:0 new windoMode:1
2021-12-01 11:04:48.385 25560-25560/com.wingkingbobo.testsnapshot W/bo.testsnapsho: Accessing hidden method Landroid/view/View;->computeFitSystemWindows(Landroid/graphics/Rect;Landroid/graphics/Rect;)Z (greylist, reflection, allowed)
2021-12-01 11:04:48.386 25560-25560/com.wingkingbobo.testsnapshot W/bo.testsnapsho: Accessing hidden method Landroid/view/ViewGroup;->makeOptionalFitsSystemWindows()V (greylist, reflection, allowed)
2021-12-01 11:04:48.392 25560-25560/com.wingkingbobo.testsnapshot W/bo.testsnapsho: Accessing hidden field Landroid/graphics/Typeface;->sSystemFontMap:Ljava/util/Map; (greylist, reflection, allowed)
2021-12-01 11:04:48.392 25560-25560/com.wingkingbobo.testsnapshot I/Mbgl-FontUtils: Couldn't map font family for local ideograph, using sans-serif instead
2021-12-01 11:04:48.399 25560-25560/com.wingkingbobo.testsnapshot I/bo.testsnapsho: Thread[1,tid=25560,Native,Thread*=0x766f810800,peer=0x747c3360,"main"] recursive attempt to load library "/data/app/com.wingkingbobo.testsnapshot-APLPZ2XgxTFTJMuFXG3bfg==/lib/arm64/libmapbox-maps.so"
2021-12-01 11:04:48.401 25560-25560/com.wingkingbobo.testsnapshot I/chatty: uid=11006(com.wingkingbobo.testsnapshot) identical 3 lines
2021-12-01 11:04:48.402 25560-25560/com.wingkingbobo.testsnapshot I/bo.testsnapsho: Thread[1,tid=25560,Native,Thread*=0x766f810800,peer=0x747c3360,"main"] recursive attempt to load library "/data/app/com.wingkingbobo.testsnapshot-APLPZ2XgxTFTJMuFXG3bfg==/lib/arm64/libmapbox-maps.so"
2021-12-01 11:04:48.415 25560-25612/com.wingkingbobo.testsnapshot I/MapboxReachability: Application permission for ACCESS_NETWORK_STATE granted
2021-12-01 11:04:48.416 25560-25620/com.wingkingbobo.testsnapshot W/CrashReporter: Root directory doesn't exist
2021-12-01 11:04:48.423 25560-25560/com.wingkingbobo.testsnapshot W/TelemetryUtils: java.lang.SecurityException: getDataNetworkTypeForSubscriber
2021-12-01 11:04:48.468 25560-25620/com.wingkingbobo.testsnapshot D/NetworkSecurityConfig: No Network Security Config specified, using platform default
2021-12-01 11:04:48.487 25560-25620/com.wingkingbobo.testsnapshot W/bo.testsnapsho: Accessing hidden field Lsun/misc/Unsafe;->theUnsafe:Lsun/misc/Unsafe; (greylist, reflection, allowed)
2021-12-01 11:04:48.488 25560-25620/com.wingkingbobo.testsnapshot W/bo.testsnapsho: Accessing hidden method Lsun/misc/Unsafe;->allocateInstance(Ljava/lang/Class;)Ljava/lang/Object; (greylist, reflection, allowed)
2021-12-01 11:04:48.542 25560-25560/com.wingkingbobo.testsnapshot D/ActivityThread: add activity client record, r= ActivityRecord{21f1c93 token=android.os.BinderProxy@9834526 {com.wingkingbobo.testsnapshot/com.wingkingbobo.testsnapshot.MainActivity}} token= android.os.BinderProxy@9834526
2021-12-01 11:04:48.545 25560-25560/com.wingkingbobo.testsnapshot I/HwPartMagicWindowFactory: add android.common.HwPartMagicWindowFactoryImpl to memory.
2021-12-01 11:04:48.550 25560-25560/com.wingkingbobo.testsnapshot I/HwPartPowerOfficeFactory: add HwPartPowerOfficeFactoryImpl to memory.
2021-12-01 11:04:48.556 25560-25560/com.wingkingbobo.testsnapshot D/InputEventReceiver: dispatchInputInterval 1000000
2021-12-01 11:04:48.571 25560-25560/com.wingkingbobo.testsnapshot I/RmeSchedManager: init Rme, version is: v1.0
2021-12-01 11:04:48.575 25560-25606/com.wingkingbobo.testsnapshot D/OpenGLRenderer: disableOutlineDraw is true
2021-12-01 11:04:48.579 25560-25604/com.wingkingbobo.testsnapshot W/LocationController: Location permissions are not granted
2021-12-01 11:04:48.580 25560-25643/com.wingkingbobo.testsnapshot D/HiTouch_PressGestureDetector: onAttached, package=com.wingkingbobo.testsnapshot, windowType=1, mHiTouchRestricted=false
2021-12-01 11:04:48.593 25560-25606/com.wingkingbobo.testsnapshot I/FrameFlow: FrameFlowInit app is not supported by frameflow solution
2021-12-01 11:04:48.593 25560-25606/com.wingkingbobo.testsnapshot I/iGraphics: [0020080c] pn: com.wingkingbobo.testsnapshot, p: 25560
2021-12-01 11:04:48.593 25560-25606/com.wingkingbobo.testsnapshot I/iGraphics: [0030080c] no spt app: com.wingkingbobo.testsnapshot
2021-12-01 11:04:48.598 25560-25610/com.wingkingbobo.testsnapshot I/FrameFlow: FrameFlowInit app is not supported by frameflow solution
2021-12-01 11:04:48.599 25560-25610/com.wingkingbobo.testsnapshot I/Mbgl-EGLConfigChooser: In emulator: false
2021-12-01 11:04:48.600 25560-25610/com.wingkingbobo.testsnapshot D/Mbgl-EglCore: EGLContext created, client version 2
2021-12-01 11:04:48.601 25560-25606/com.wingkingbobo.testsnapshot D/mali_winsys: EGLint new_window_surface(egl_winsys_display *, void *, EGLSurface, EGLConfig, egl_winsys_surface **, EGLBoolean) returns 0x3000
2021-12-01 11:04:48.601 25560-25610/com.wingkingbobo.testsnapshot D/mali_winsys: EGLint new_window_surface(egl_winsys_display *, void *, EGLSurface, EGLConfig, egl_winsys_surface **, EGLBoolean) returns 0x3000
2021-12-01 11:04:48.601 25560-25610/com.wingkingbobo.testsnapshot I/maps: Using OpenGL render backend
2021-12-01 11:04:48.608 25560-25606/com.wingkingbobo.testsnapshot W/Gralloc3: mapper 3.x is not supported
2021-12-01 11:04:48.612 25560-25660/com.wingkingbobo.testsnapshot I/maps: {MapboxRenderThr}[General]: GPU Identifier: Mali-G76
2021-12-01 11:04:48.636 25560-25560/com.wingkingbobo.testsnapshot I/HwViewRootImpl: removeInvalidNode jank list is null
2021-12-01 11:04:48.638 25560-25560/com.wingkingbobo.testsnapshot D/DecorView: showOrHideHighlightView: hasFocus=true; winMode=1; isMrgNull=true
2021-12-01 11:04:48.639 25560-25560/com.wingkingbobo.testsnapshot W/InputMethodManager: startInputReason = 1
2021-12-01 11:04:48.652 25560-25560/com.wingkingbobo.testsnapshot W/InputMethodManager: startInputReason = 5
2021-12-01 11:04:48.753 25560-25612/com.wingkingbobo.testsnapshot D/HwCustConnectivityManagerImpl: isBlockNetworkRequestByNonAis, INVALID_SUBSCRIPTION_ID
2021-12-01 11:04:51.052 25560-25560/com.wingkingbobo.testsnapshot I/Hwaps: APS: EventAnalyzed: initAPS: version is 11.0.0.4
2021-12-01 11:04:51.053 25560-25560/com.wingkingbobo.testsnapshot D/Hwaps: Fpsrequest create,type:EXACTLY_IDENTIFY
2021-12-01 11:04:51.053 25560-25560/com.wingkingbobo.testsnapshot D/Hwaps: Fpsrequest create,type:EXACTLY_IDENTIFY
2021-12-01 11:04:51.053 25560-25560/com.wingkingbobo.testsnapshot D/Hwaps: Fpsrequest create,type:OPENGL_SETTING
2021-12-01 11:04:51.054 25560-25560/com.wingkingbobo.testsnapshot D/Hwaps: FpsController create
2021-12-01 11:04:51.057 25560-25560/com.wingkingbobo.testsnapshot D/Hwaps: APS: EventAnalyzed: reInitFpsPara :mBaseFps = 60; mMaxFps = 60
2021-12-01 11:04:51.059 25560-25560/com.wingkingbobo.testsnapshot W/Settings: Setting device_provisioned has moved from android.provider.Settings.Secure to android.provider.Settings.Global.
2021-12-01 11:04:51.060 25560-25560/com.wingkingbobo.testsnapshot V/HiTouch_HiTouchSensor: User setup is finished.
2021-12-01 11:04:51.125 25560-25560/com.wingkingbobo.testsnapshot W/HwApsManager: HwApsManagerService, registerCallback, start !
2021-12-01 11:04:51.126 25560-25560/com.wingkingbobo.testsnapshot D/Hwaps: APS: EventAnalyzed: registerCallbackInApsManagerService, mPkgName:com.wingkingbobo.testsnapshot; result = true
2021-12-01 11:04:51.129 25560-25560/com.wingkingbobo.testsnapshot V/AudioManager: querySoundEffectsEnabled...
2021-12-01 11:04:51.203 25560-25610/com.wingkingbobo.testsnapshot D/HwGalleryCacheManagerImpl: mIsEffect:false
2021-12-01 11:04:53.295 25560-25560/com.wingkingbobo.testsnapshot I/AwareBitmapCacher: init lrucache size: 2097152 pid=25560

```

**Comment by kiryldz** [comment]
@wingkingbobo sadly I don't have any of such Huawei phones nearby.
Can you try performing snapshot in your example with Huawei with following function? It will produce flipped image but just to check it will be valid.
```
  private fun testSnapshot(mapView: MapView, callback: (Bitmap?) -> Unit) {
    mapView.queueEvent(
      {
        val buffer = ByteBuffer
          .allocateDirect(mapView.width * mapView.height * 4)
          .order(ByteOrder.nativeOrder())
        GLES20.glReadPixels(
          0, 0, mapView.width, mapView.height,
          GLES20.GL_RGBA, GLES20.GL_UNSIGNED_BYTE, buffer
        )
        buffer.rewind()
        runOnUiThread {
          val bmp = Bitmap.createBitmap(
            mapView.width,
            mapView.height,
            Bitmap.Config.ARGB_8888
          ).apply {
            copyPixelsFromBuffer(buffer)
          }
          callback.invoke(bmp)
        }
      },
      needRender = false
    )
  }
```
and call it in your example:
```
class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        map_view.getMapboxMap().loadStyleUri(Style.SATELLITE_STREETS) {
            btn_snapshot.setOnClickListener {

                testSnapshot(map_view) { bmp ->
                    image_view.setImageBitmap(bmp)
                }
            }
        }
    }
}
```

**Comment by wingkingbobo** [comment]
@kiryldz On honor 20 pro, testsnapshot returns a black image, while on redmi note 9, it is a flipped image.

**Comment by kiryldz** [comment]
> @kiryldz On honor 20 pro, testsnapshot returns a black image, while on redmi note 9, it is a flipped image.

Thanks @wingkingbobo for helping out with testing!
We identified the root cause, fix will be delivered in Jan 2022 as part of `v10.3-beta.1` release. Will keep you updated!

**Comment by yunikkk** [comment]
@wingkingbobo fix is merged and will arrive in 10.3 release

**Comment by wingkingbobo** [comment]
@yunikkk Ok, Thanks
